module/zcommon/zfs_fletcher_intel.c

```diff
@@ -0,0 +1,150 @@
+/*
+ * Implement fast Fletcher4 with AVX2 instructions. (x86_64)
+ *
+ * Use the 256-bit AVX2 SIMD instructions and registers to compute
+ * Fletcher4 in four incremental 64-bit parallel accumulator streams,
+ * and then combine the streams to form the final four checksum words.
+ *
+ * Copyright (C) 2015 Intel Corporation.
+ *
+ * Authors:
+ *      James Guilford <james.guilford@intel.com>
+ *      Jinshan Xiong <jinshan.xiong@intel.com>
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * OpenIB.org BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#if defined(HAVE_AVX) && defined(HAVE_AVX2)
+
+#include <linux/simd_x86.h>
+#include <sys/spa_checksum.h>
+#include <zfs_fletcher.h>
+
+static void
+fletcher_4_avx2_init(zio_cksum_t *zcp)
+{
+	ZIO_SET_CHECKSUM(zcp, 0, 0, 0, 0);
+
+	kfpu_begin();
+
+	/* clear avx2 registers */
+	asm volatile("vpxor %ymm0, %ymm0, %ymm0");
+	asm volatile("vpxor %ymm1, %ymm1, %ymm1");
+	asm volatile("vpxor %ymm2, %ymm2, %ymm2");
+	asm volatile("vpxor %ymm3, %ymm3, %ymm3");
```

Comment:
Technically, it is wrong to assume that ymm regs will not change across function
calls. This is especially true when `kfpu_begin()` is a noop (userspace, OSX kernel, etc...)

This can be avoided in the following way:

```
union fletcher4_cookie {
#if defined(HAVE_AVX) && defined(HAVE_AVX2)
	uint64_t abcd[4*4] __attribute__((aligned(32)));
#endif
}

fletcher_4_native(zcp) {
	fletcher4_cookie c;
	init(&c); // bzero(c)
	compute(buf, &c); // load(c->ymm), vect_cal, store(ymm->c)
	fini(zcp, &c); // use c to get CHECKSUM
}
```